plugin/src/main/java/org/opensearch/ml/action/prediction/TransportPredictionTaskAction.java

```diff
@@ -180,6 +186,67 @@ public void onFailure(Exception e) {
         }
     }
 
+    private void deployAndPredictRemoteModel(
+        String modelId,
+        ActionListener<MLTaskResponse> wrappedListener,
+        MLPredictionTaskRequest mlPredictionTaskRequest
+    ) {
+        String[] workerNodes = mlModelManager.getWorkerNodes(modelId, FunctionName.REMOTE, true);
+        if (workerNodes != null && workerNodes.length != 0) {
+            // transport layer deployment finished
+            executePredict(mlPredictionTaskRequest, wrappedListener, modelId);
+            return;
+        }
+        mlModelManager.getModel(modelId, ActionListener.wrap(mlModel -> {
+            // Obtain the deployment plan
+            String[] workerNodeIds = mlModel.getPlanningWorkerNodes();
+            MLDeployModelRequest deployModelRequest = MLDeployModelRequest
+                .builder()
+                .modelId(modelId)
+                .async(false)
+                .dispatchTask(false)
+                .modelNodeIds(workerNodeIds)
+                .build();
+            deployAndPredict(modelId, deployModelRequest, mlPredictionTaskRequest, wrappedListener);
+        }, exception -> {
+            log.error("Failed to find the remote model in the model index " + modelId, exception);
+            wrappedListener.onFailure(exception);
+        }));
+    }
+
+    private void deployAndPredict(
+        String modelId,
+        MLDeployModelRequest deployModelRequest,
+        MLPredictionTaskRequest mlPredictionTaskRequest,
+        ActionListener<MLTaskResponse> wrappedListener
+    ) {
+        ActionListener<MLDeployModelResponse> deployModelActionListener = ActionListener.wrap(deployModelResponse -> {
+            // Deployment failed, existing
+            if (!deployModelResponse.getStatus().equals(MLTaskState.COMPLETED.name())) {
+                wrappedListener
+                    .onFailure(
+                        new IllegalArgumentException(
+                            "Deployment Failed for " + modelId + ", please check your remote model before predicting."
+                        )
+                    );
+                return;
+            }
+            String[] workerNodes = mlModelManager.getWorkerNodes(modelId, FunctionName.REMOTE, true);
+            if (workerNodes == null || workerNodes.length == 0) {
+                // The deployment is kicked off, but not finished yet in this node
+                mlModelManager.deployRemoteModelToLocal(modelId, ActionListener.wrap(r -> {
```

Comment:
So we will deploy again to local here , will it cause any issue/conflict with the first deployment in line 247?

Reply:
Good question. I have been thinking this too. My current observation is that usually the local deploy wins and invoke the model, and later the transport layer deploy arrives and overwrite. I haven't implemented bypassing modelCache updates when a value already exists from an earlier deploy (likely not necessary).

For remote models, here are the two variables that to be updated in the deployment. 
1. Predictable predictor
2. Set<String> workerNodes

The workerNodes is initialized as ConcurrentHashMap.newKeySet() so it's thread safe. The predictor is a Predictor value and it's updated through "public synchronized void setPredictor()" and it's also thread safe. 

Basically, during the short window (xx ms) between the cluster deploy and local deploy, we will always "predict" a model based on the latest local deployment value. Once the cluster level deployment is finished, the "predict" will not deploy anymore.